### PR TITLE
OOP school library: unit tests

### PR DESCRIPTION
**Pull request by Salwa99**
# OOP school library: unit tests

### Howdy! :cowboy_hat_face: 

**In this PR We :**

1. Write unit tests using `Rspec` for the following classes : 

    - `Book`
    - `Person`
    - `Student`
    - `Teacher`
    - `Classroom`
    - `Rental`

2. Follow GitFlow.
3. Fix Linter errors.

_Cheers! :rose:_

### CODE REVIEW

File: spec/person_spec.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  it 'Check if the person can use services' do
+    expect(person.can_use_services?).to be_truthy
+  end
+
+  it 'Check the rentals of the person' do
+    expect(person.rentals).to be_empty
+  end
+end
```
Comment (VICTOR-CHUKWU) [inline spec/person_spec.rb:25]: - [ ] you should try and test a good number of the features in the person class e. test to show that rental is added to person,  test to show `The constructor method sets default value for the optional name parameter if it is not provided`, I can not see test to validate correct name. kindly add it too
```
it "The add_rental method returns a rental and adds its to the book" do
      book = Book.new("Title", "Author")
      person = Person.new(24, "Bob")

      rental = person.add_rental(book, "2021-05-04")

      expect(person.rentals).to eql [rental]
    end
```

File: spec/book_spec.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  describe '#author' do
+    it 'Return the book author' do
+      expect(book.author).to eq 'Friedrich Nietzsche'
+    end
+  end
+
+  describe '#rentals' do
+    it 'Return an empty array' do
+      expect(book.rentals).to be_empty
+    end
```
Comment (VICTOR-CHUKWU) [inline spec/book_spec.rb:27]: - [ ] kindly add a test to show that rental is added to a book

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi Team 👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good PR
- Good commit
- All linters passed


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by bimbolabuari — COMMENTED** [review]
Hello Team 👋,

_This is an additional review as a result of the second opinion request._

- Kudos on working on this milestone 👍 
    - As part of the project requirement:
       - There is a requirement to add unit tests for all the methods in all the classes(`Book`, `Person`, `Student`, `Teacher`, `Classroom`, `Rental`:

- In your `book.rb`, there is an add rental method that was not tested.
  - https://github.com/AmanWorku/OOP-School-Library/blob/87561ee551364513ecce9bf407cce85296754a2d/book.rb#L8-L10

- Also in your `person.rb`, there are some method that were not tested as required.
 - https://github.com/AmanWorku/OOP-School-Library/blob/87561ee551364513ecce9bf407cce85296754a2d/person.rb#L23-L25
 - https://github.com/AmanWorku/OOP-School-Library/blob/87561ee551364513ecce9bf407cce85296754a2d/person.rb#L27-L29
 - https://github.com/AmanWorku/OOP-School-Library/blob/87561ee551364513ecce9bf407cce85296754a2d/person.rb#L31-L39
 - https://github.com/AmanWorku/OOP-School-Library/blob/87561ee551364513ecce9bf407cce85296754a2d/person.rb#L41-L43
 - https://github.com/AmanWorku/OOP-School-Library/blob/87561ee551364513ecce9bf407cce85296754a2d/person.rb#L19-L21

- Currently, you instantiate the classes ("Note: The phrase "instantiating a class" means the same thing as "creating an object." When you create an object, you are creating an "instance" of a class, therefore "instantiating" a class.") and test for the new object variable 

- Let's check this [previous exercise link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/lessons/rspec_unit_test.md#unit-testing) for more knowledge of unit testing and please let's make the requested changes asked by the previous code reviewer. 👍 


Keep rocking 🚀!

File: spec/rental_spec.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+RSpec.describe Rental do
+  date = '22/03/2023'
+  book = Book.new('Watchmen', 'Ala Moore')
+  person = Person.new(25, 'Aman')
+  subject(:rental) { described_class.new(date, book, person) }
+
+  describe '#date' do
+    it 'returns the correct date' do
+      expect(rental.date).to eq '22/03/2023'
+    end
```
Comment (frank1738) [inline spec/rental_spec.rb:14]: - [ ] Good work adding tests but some tests are failing. Kindly check them out

image

**Review by frank1738 — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

✅ Correct use of Gitflow
✅ No linter error



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

N/A
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/rental_spec.rb (hunk 3 above)
Comment (Salwa99) [reply spec/rental_spec.rb:14]: Hi @frank1738 , I can't see any errors in out tests
Screenshot from 2023-03-23 16-01-52

**Review by addod19 — APPROVED** [review]
### Hi @AmanWorku and @Salwa99, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

### Highlights
- [ ] All test cases passed :rocket: 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by AmanWorku** [comment]
Thank you @addod19.

**Pull request merged**